src/java/org/apache/cassandra/index/sai/utils/RangeIntersectionIterator.java

```diff
@@ -0,0 +1,229 @@
+ */
+@SuppressWarnings("resource")
+public class RangeIntersectionIterator extends RangeIterator
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    // The cassandra.sai.intersection.clause.limit (default: 2) controls the maximum number of range iterator that
+    // will be used in the final intersection of a query operation.
+    private static final int INTERSECTION_CLAUSE_LIMIT = Integer.getInteger("cassandra.sai.intersection.clause.limit", 2);
+
+    static
+    {
+        logger.info(String.format("Storage attached index intersection clause limit is %d", INTERSECTION_CLAUSE_LIMIT));
+    }
+
+    private final List<RangeIterator> ranges;
+
+    private RangeIntersectionIterator(Builder.Statistics statistics, List<RangeIterator> ranges)
+    {
+        super(statistics);
+        this.ranges = ranges;
+    }
+
+    protected PrimaryKey computeNext()
+    {
+        // Range iterator that has been advanced in the previous cycle of the outer loop.
+        // Initially there hasn't been the previous cycle, so set to null.
+        int alreadyAvanced = -1;
+
+        // The highest primary key seen on any range iterator so far.
+        // It can become null when we reach the end of the iterator.
+        PrimaryKey highestKey = getCurrent();
+
+        outer:
+        while (highestKey != null && highestKey.compareTo(getMaximum()) <= 0)
+        {
+            // Try advance all iterators to the highest key seen so far.
+            // Once this inner loop finishes normally, all iterators are guaranteed to be at the same value.
+            for (int index = 0; index < ranges.size(); index++)
+            {
+                if (index != alreadyAvanced)
+                {
+                    RangeIterator range = ranges.get(index);
+                    PrimaryKey nextKey = nextOrNull(range, highestKey);
+                    if (nextKey == null || nextKey.compareTo(highestKey) > 0)
+                    {
+                        // We jumped over the highest key seen so far, so make it the new highest key.
+                        highestKey = nextKey;
+                        // Remember this iterator to avoid advancing it again, because it is already at the highest key
+                        alreadyAvanced = index;
+                        // This iterator jumped over, so the other iterators are lagging behind now,
+                        // including the ones already advanced in the earlier cycles of the inner loop.
+                        // Therefore, restart the inner loop in order to advance
+                        // the other iterators except this one to match the new highest key.
+                        continue outer;
+                    }
+                    assert nextKey.compareTo(highestKey) == 0:
+                    String.format("skipTo skipped to an item smaller than the target; " +
```

Comment:
nit: 
```suggestion
                    String.format("nextOrNull() skipped to an item smaller than the target; " +
```
...or could just remove that and say just "skipped to" in case the method changes, etc.